Configure goreleaser to build client and server

torstenwalter opened the pull request:
We noticed that this project does not have any releases. It would be great if versions could be tagged and binaries be provided.
This PR demonstrates how this can be done using GitHub Actions and [goreleaser](https://goreleaser.com/).

Once this is merged all one need to do is to create a tag in GitHub following the naming convention "v?.?.?" and a release will be created automatically. You can check https://github.com/syncier/aws-encryption-provider/releases/tag/v0.0.1. It was created that way. As you can see it also contains a changelog.

Things this PR does:

- add github workflows for ci and release
- Adjusted ldflags to match Makefile
- formatted files with go fmt
- go mod tidy

torstenwalter commented:
Could you have a look at this?

micahhausler commented on .goreleaser.yml:
```diff
@@ -0,0 +1,45 @@
+      - linux
+      - windows
+      - darwin
+    ldflags: -s -w -X sigs.k8s.io/aws-encryption-provider/pkg/version.Version={{.Version}} -X sigs.k8s.io/aws-encryption-provider/pkg/version.Commit={{.Commit}} -X sigs.k8s.io/aws-encryption-provider/pkg/version.Date={{.Date}}
```
Can you add a `Commit` and `Date` variable in `pkg/version/version.go`?

micahhausler commented on .goreleaser.yml:
```diff
@@ -0,0 +1,45 @@
+      - windows
+      - darwin
+    ldflags: -s -w -X sigs.k8s.io/aws-encryption-provider/pkg/version.Version={{.Version}} -X sigs.k8s.io/aws-encryption-provider/pkg/version.Commit={{.Commit}} -X sigs.k8s.io/aws-encryption-provider/pkg/version.Date={{.Date}}
+  - id: grpcclient
```
The client is purely for testing, I don't think it makes sense to include it release artifacts

micahhausler commented on .goreleaser.yml:
```diff
@@ -0,0 +1,45 @@
+    env:
+      - CGO_ENABLED=0
+    goos:
+      - linux
```
Linux is the only supported kube-apiserver target, so we don't need a windows or darwin build

micahhausler commented on .goreleaser.yml:
```diff
@@ -0,0 +1,45 @@
+# This is an example goreleaser.yaml file with some sane defaults.
+# Make sure to check the documentation at http://goreleaser.com
+before:
```
I don't think we need anything in this `before` block since we vendor dependencies and don't generate code

micahhausler commented on .goreleaser.yml:
```diff
@@ -0,0 +1,45 @@
+    - go generate ./...
+builds:
+  - id: grpcserver
+    binary: grpcserver
```
Can you change this for consistency with the package name? Can you also update the `Makefile` to use the package name for the server binary?
```suggestion
    binary: aws-encryption-provider
```

micahhausler commented on .goreleaser.yml:
```diff
@@ -0,0 +1,45 @@
+      - linux
+      - windows
+      - darwin
+archives:
```
Do we need any of these set? I think we can safely remove this block

micahhausler commented on .goreleaser.yml:
```diff
@@ -0,0 +1,45 @@
+checksum:
+  name_template: 'checksums.txt'
+snapshot:
+  name_template: "{{ .Tag }}-next"
```
I don't think we need anything in this block that is different than the default value. Can you drop it?

micahhausler commented on .goreleaser.yml:
```diff
@@ -0,0 +1,45 @@
+  name_template: "{{ .Tag }}-next"
+changelog:
+  sort: asc
+  filters:
```
Can you remove the filters? We don't enforce structured commit messages

micahhausler commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,32 @@
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
+      - name: Ensure gofmt
+        run: test -z "$(gofmt -s -d cmd pkg)"
+      - name: Ensure go.mod is already tidied
+        run: go mod tidy && git diff --no-patch --exit-code
+      - name: Run unit tests
+        run: go test ./...
```
I like the idea of tidying the go modules and checking for gofmt, but can you put those in the `Makefile`? We already have [Prow doing the tests](https://github.com/kubernetes/test-infra/blob/f34ec251f8b943950bbd0718e13c47b661ea97da/config/jobs/kubernetes-sigs/aws-encryption-provider/presubmits.yaml), and I want to keep the CI similar to the dev workflow (just `make test`).

I think running goreleaser with `--snapshot --skip-publish --rm-dist` on PRs in GH workflow is reasonable

micahhausler commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,32 @@
+      - name: Setup Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.14
```
I'm fine bumping the project to use 1.14 or even 1.15, but we'll need to update `go.mod` and the [Prow presubmits](https://github.com/kubernetes/test-infra/blob/f34ec251f8b943950bbd0718e13c47b661ea97da/config/jobs/kubernetes-sigs/aws-encryption-provider/presubmits.yaml) container images

torstenwalter replied:
Sure PR for prow jobs: https://github.com/kubernetes/test-infra/pull/20702

micahhausler reviewed (changes requested):
Thank you for this PR, and sorry it hasn't been reviewed sooner! 

/ok-to-test

torstenwalter commented:
@micahhausler Thank you for the review and all the suggestions. I updated as suggested. `gofmt` and `go mod tidy` now have small scripts in hack to stay in line with the existing code.

micahhausler commented:
This looks great! Can you update `go.mod` to reference 1.15 as well?

torstenwalter commented:
Sure. I missed that one.

torstenwalter commented:
The last commit became bigger. The issue is that ci downloads` golangci-lint` and adds it to `go.mod`. After that the dependencies are no longer vendored. I now added `golangci-lint` directly to `go.mod`. To ensure that it's not removed by `go mod tidy` I added tools.go, which imports it.

micahhausler commented:
/lgtm 
/approve

torstenwalter commented:
@micahhausler Thank you for merging this!
Could you create and push a new tag "v*.*.*" to trigger a release so that we see if everything is working?
https://github.com/kubernetes-sigs/aws-encryption-provider/blob/master/.github/workflows/release.yaml#L5
